cluster/src/main/java/org/apache/iotdb/cluster/log/manage/serializable/SyncLogDequeSerializer.java

```diff
@@ -50,82 +53,140 @@
-  private Deque<Integer> logSizeDeque = new ArrayDeque<>();
+  private FileOutputStream currentLogDataOutputStream;
+  private FileOutputStream currentLogIndexOutputStream;
   private LogManagerMeta meta;
   private HardState state;
-  // mark first log position
-  private long firstLogPosition = 0;
-  // removed log size
-  private long removedLogSize = 0;
-  // when the removedLogSize larger than this, we actually delete logs
-  private long maxRemovedLogSize = ClusterDescriptor.getInstance().getConfig()
-      .getMaxUnsnapshotLogSize();
-  // min version of available log
+
+  /**
+   * min version of available log
+   */
   private long minAvailableVersion = 0;
-  // max version of available log
+
+  /**
+   * max version of available log
+   */
   private long maxAvailableVersion = Long.MAX_VALUE;
-  // log dir
+
   private String logDir;
-  // version controller
+
   private VersionController versionController;
 
-  private ByteBuffer logBuffer = ByteBuffer
+  private ByteBuffer logDataBuffer = ByteBuffer
       .allocate(ClusterDescriptor.getInstance().getConfig().getRaftLogBufferSize());
+  private ByteBuffer logIndexBuffer = ByteBuffer
+      .allocate(ClusterDescriptor.getInstance().getConfig().getRaftLogBufferSize());
+
+  private long offsetOfTheCurrentLogDataOutputStream = 0;
+
+  /**
+   * file name pattern:
+   * <p>
+   * for log data file: ${startTime}-${Long.MAX_VALUE}-{version}-data
+   * <p>
+   * for log index file: ${startTime}-${Long.MAX_VALUE}-{version}-idx
+   */
+  private static final int FILE_NAME_PART_LENGTH = 4;
 
-  private final int flushRaftLogThreshold = ClusterDescriptor.getInstance().getConfig()
-      .getFlushRaftLogThreshold();
+  private int maxRaftLogIndexSizeInMemory = ClusterDescriptor.getInstance().getConfig()
+      .getMaxRaftLogIndexSizeInMemory();
 
-  private int bufferedLogNum = 0;
+  private int maxRaftLogPersistDataSizePerFile = ClusterDescriptor.getInstance().getConfig()
+      .getMaxRaftLogPersistDataSizePerFile();
 
+  private int maxNumberOfPersistRaftLogFiles = ClusterDescriptor.getInstance().getConfig()
+      .getMaxNumberOfPersistRaftLogFiles();
+
+  private int maxPersistRaftLogNumberOnDisk = ClusterDescriptor.getInstance().getConfig()
+      .getMaxPersistRaftLogNumberOnDisk();
+
+  private ScheduledExecutorService persistLogDeleteExecutorService;
+  private ScheduledFuture<?> persistLogDeleteLogFuture;
+
+  /**
+   * indicate the first raft log's index of {@link SyncLogDequeSerializer#logIndexOffsetList}, for
+   * example, if firstLogIndex=1000, then the offset of the log index 1000 equals
+   * logIndexOffsetList[0], the offset of the log index 1001 equals logIndexOffsetList[1], and so
+   * on.
+   */
+  private long firstLogIndex = 0;
+
+  /**
+   * the offset of the log's index, for example, the first value is the offset of index
+   * ${firstLogIndex}, the second value is the offset of index ${firstLogIndex+1}
+   */
+  private List<Long> logIndexOffsetList;
```

Comment:
As currently the  default maxSize of each data file is 1G,it's seems a Integer is able to record the offset.

Reply:
sure, but the parameter of the max size of the log data file is left to the user to set, so if the user config larger than an integer, it would occur errors, or we could limit the range of user configuration for the maxSize parameter.

Reply:
OK~Just let it go~Long is OK for me ~